src/components/ProteinViewer/utils.ts

```diff
@@ -1,7 +1,60 @@
 import { createSelector } from 'reselect';
 import { toPlural } from 'utils/pages/toPlural';
 import { NOT_MEMBER_DBS } from 'menuConfig';
+import { getTrackColor, EntryColorMode } from 'utils/entry-color';
 
+const dbs4SingleDomain = [
+  'pfam',
+  'smart',
+  'pirsf',
+  'ncbifam',
+  'hamap',
+  'sfld',
+  'cdd',
+  'profile',
+];
+
+const selectRepresentativeDomains = (domains: Record<string, unknown>[]) => {
+  const flatDomains = [];
+  for (const domain of domains) {
+    const { accession, short_name, source_database, integrated } = domain;
+    for (const location of domain.entry_protein_locations as Array<ProtVistaLocation>) {
+      for (const fragment of location.fragments) {
+        const { start, end } = fragment;
+        flatDomains.push({
+          accession,
+          short_name,
+          source_database,
+          integrated,
+          start,
+          end,
+          color: getTrackColor({ source_database }, EntryColorMode.MEMBER_DB),
+          length: end - start,
+          keep: true,
+        });
+      }
+    }
+  }
+  // const sortedDomains = flatDomains.sort((a, b) => b.length - a.length);
+  for (const dom1 of flatDomains) {
+    for (const dom2 of flatDomains) {
+      if (dom1 === dom2 || !dom1.keep || !dom2.keep) continue;
```

Comment:
Let N be the number of domains in `flatDomains`.

This code will perform at most $(N^2 - N)$ comparisons, but $(N^2 - N) \over 2$ would be enough.

```ts
flatDomains.forEach((dom1, i) => {
  flatDomains.slice(i+1).forEach((dom2) => {
    ...
  });
});
```